Grand Abalone

r-gke opened the pull request:
This pull request is made to insert a new variant of Abalone called 'Grand Abalone'. It is a very simply defined:
1. a larger board;
2. more marbles can be used in a single action (4 instead of 3);
3. you win by ejecting 10 marbles instead of 6;
4. each player plays two moves during their turn, except for player 1 at the start, who only gets one move.

The rest is identical to standard Abalone.

Despite this simplicity, the current code for Abalone is not written in a way that makes such extensions easy to implement (if we want to avoid duplications). Therefore, I rewrote some parts of it accordingly, before introducing the new variant, which is now achieved in fewer than 50 lines of code.

msomatt reviewed (changes requested):
Hi Alex,

Firstly, thanks for taking the time to add Grand Abalone. You've made a good job of it overall.

I've spent the last couple of hours trying to pick through the full review. You've made that task more difficult because there are stylistic changes you've made that just add noise to the diff. In some cases I've asked for you to revert them so as to keep consistency throughout the strategygames codebase.

I've done my review by eye, as I've not currently got a dev environment setup (old laptop died, new laptop on the way). So I've not even looked at the tests, and assume they all pass, but i've not run them.

I acknowledge you have completely rewritten the move generation, which looks to be more scalable/sustainable across Abalone and Grand Abalone. I'll probably review this fully another day.

I started looking at the changes to the `Pos` file, which also looked pretty sweeping, and I'll want to think about these some more too. We need to ensure that old Abalone games load out of the database ok.

Have a look at my comments and submit I'll be in a better place to review more thoroughly next week with a new laptop.

msomatt commented:
Can you also explain the `Norm` module? Why are there 4 Norm types (N4, N6, N8, N12)?

r-gke commented:
> Can you also explain the `Norm` module? Why are there 4 Norm types (N4, N6, N8, N12)?

Hi, thanks for your time! Only N6 is actually used, the other ones only stand for potential other variants. The norm is here to describe the topology of the board, in particular: the neighbourhood of a given position, the directions of movement; and (in the case of Abalone) the shape of the whole board. Then, if you want to create a variant of Abalone on, say, a square grid (e.g. Traboulet), you just have to change the norm (e.g. use N4 instead of N6).

r-gke commented on src/main/scala/abalone/variant/GrandAbalone.scala:
```diff
@@ -0,0 +1,43 @@
+      id = 2,
+      key = "grand_abalone",
+      name = "Grand Abalone",
+      standardInitialPosition = true,
```
What is the meaning of this boolean?

msomatt replied:
I think its a boolean stating whether the initial fen for the variant matches the initial fen of the default variant. But its not used a lot, its lichess historic

r-gke commented on src/main/scala/Move.scala:
```diff
@@ -353,14 +337,13 @@ object Move {
   final case class Abalone(m: abalone.Move)
       extends Move(
-        Piece.Abalone(m.piece),
+        Piece.Abalone(m.situationBefore.board.getPiece(m.orig).get),
```
Indeed, it may go awry if the move is illegal, but what shall I return if there is no piece? (remark: I removed the piece from the Abalone moves because it is irrelevant to their mechanics)

msomatt replied:
I guess you can just construct the Piece out of the Player (which we have in the Abalone Move) and the Role (which is always just Stone which is the defaultRole for abalone)

r-gke commented on src/main/scala/abalone/Board.scala:
```diff
@@ -1,33 +1,24 @@
-  def isEmptySquare(pos: Option[Pos]): Boolean = pos.fold(false)(!this.pieces.contains(_))
+  def isPiece(a: Pos): Boolean = pieces.contains(a)
+
+  def getPiece(a: Pos): Option[Piece] = pieces.get(a)
```
I'll as you say, but I still think it better describes what it does as I wrote it.

r-gke commented on src/main/scala/abalone/Hash.scala:
```diff
@@ -17,49 +16,50 @@ final class Hash(size: Int) {
+    * index is < posNb.
+    */
+  def actorIndex(sit: Situation, actor: Actor) =
+    sit.board.variant.boardType.posNb * actor.piece.player.fold(0, 1) + actor.pos.hashIndex
```
OK. Seems to me rather arbitrary, anyway.

r-gke commented on src/main/scala/abalone/Pos.scala:
```diff
@@ -1,525 +1,400 @@
-    // Pos.J8.index  -> '¦'
-  )
-}
+  def add(x: Int, y: Int): Pos = {
```
Clearly, that's not what I wanted, I have no issue with my objects being mutable. I'll change them anyway.

r-gke commented on src/main/scala/abalone/Pos.scala:
```diff
@@ -1,525 +1,400 @@
 package strategygames.abalone
 
-/*
```
Sorry for that...

r-gke commented on src/main/scala/abalone/format/FEN.scala:
```diff
@@ -1,25 +1,24 @@
 
-  def pieces: PieceMap = value
+  // Notice cells are described from bottom left to top right in the FEN
+  def pieces(boardType: BoardType): PieceMap = value
```
Since the FEN does not store the variant, I think it would be even better to have the variant in parameter. Determining the board from the FEN is possible here, but looks like a hack.
What is the issue in passing a parameter through this function?

msomatt replied:
A lot of the time when you have a FEN you have a Variant as well. In the chess gamelogic, we have 10 different chess variants, and they all use the same FEN functions agnostic of what variant it is.

A stated originally, counting the number of `/` in the fen and that should tell you how many rows there are? Is there a problem with this approach?

r-gke commented on src/main/scala/abalone/format/Uci.scala:
```diff
@@ -1,52 +1,43 @@
     def apply(move: String): Option[Move] =
       move match {
-        case moveR(orig, dest) =>
+        case moveR(orig0, orig1, dest0, dest1) =>
```
It happened because the parentheses are no longer the same in the regex.

r-gke commented on src/main/scala/abalone/format/pgn/Binary.scala:
```diff
@@ -40,38 +39,37 @@ object Binary {
+    def moveUci(b1: Int, b2: Int): String = s"${posFromInt(b1)}${posFromInt(b2)}"
 
-    def posFromInt(b: Int): String = Pos(right(b, 7)).get.toString()
+    def posFromInt(b: Int): String = Pos.fromIndex(right(b, 7)).toString()
```
The Piotr I used features the same indices than the one for Go, and the same indices as previously defined for Abalone. Unfortunately, since the order is not really straightforward, that means the index can go beyond 120 (I think 199).

msomatt replied:
Ok if its beyond 128 then we need to use all 8 bits then, so this needs updating. This isn't tied to piotr btw.

r-gke commented on src/main/scala/abalone/variant/Abalone.scala:
```diff
@@ -1,23 +1,22 @@
-  override def baseVariant: Boolean =
-    true // Belgian Daisy initialFen is defined in Abalone default "Variant" file
+  /** Belgian daisy. */
+  override def initialFen: FEN = FEN("SS1ss/SSSsss/1SS1ss1/8/9/8/1ss1SS1/sssSSS/ss1SS 0 0 b 0 1")
```
I did not reverse the case, I reversed the rows, to start from the bottom left rather the the top left. Does that raise an issue with the existing data?

msomatt replied:
Yeah, sorry we need to stick with the original format as this will break all current Abalone games in the database. Whether you agree with it or not, the FEN format is pretty well adopted for games (originally Chess) and its starts top to bottom. Its not a lichess/playstrategy thing, its much more widely used than that: https://en.wikipedia.org/wiki/Forsyth%E2%80%93Edwards_Notation

r-gke replied:
Arg, Vincent had told me it should be OK... I'll have to rewrite all FENs in the code.

So, it means the games are stored in the database as lists of FENs?

---
As for why I dod not agree with it (I really do not mind whether it is widely used or not): I understand we read from top to bottom to mimic a written page, but it strikes me as inconsistent to read the columns in the order they are numbered (from left to right) and not do the same for the rows (from bottom to top).

r-gke commented on src/main/scala/abalone/variant/GrandAbalone.scala:
```diff
@@ -0,0 +1,43 @@
+case object GrandAbalone
+    extends Variant(
+      id = 2,
+      key = "grand_abalone",
```
OK.

r-gke commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -391,17 +370,20 @@ abstract class Variant private[variant] (
   def exists(id: Int): Boolean = byId contains id
 
-  val openingSensibleVariants: Set[Variant] = Set(strategygames.abalone.variant.Abalone)
+  val openingSensibleVariants: Set[Variant] = Set(Abalone, GrandAbalone)
```
What is the meaning of `openingSensibleVariants`?

msomatt replied:
I think its whether the opening book moves make sense for this variant. Again its all lichess historic and we haven't done loads with it, but given the opening book moves work off FENs and the FENs dont match for Abalone and GrandAbalone, then GrandAbalone should be left out here

r-gke commented on src/main/scala/abalone/Game.scala:
```diff
@@ -75,21 +75,20 @@ case class Game(
           situation = Situation(
-            board = parsed.situation.board withVariant g.board.variant,
-            player = parsed.situation.player
+            board = parsed.sit.board withVariant g.board.variant,
```
OK.

r-gke commented on src/main/scala/abalone/History.scala:
```diff
@@ -1,38 +1,44 @@
     currentTurn: List[Uci] = List.empty,
+    // Contained in lastMove, but we may want to know the previous player without knowing the previous move
+    prevPlayer: Option[Player] = None,
+    prevMove: Option[LightMove] = None,// Remark: putting the actual move here slows down everything past a few turns, because of the previous and next situations contained within the move
```
I'm sorry, I don't really understand the need of having a 'history' instead of
1. the last action (Uci);
2. a pointer towards the previous situation.
Then I would save the game by storing the initial situation and the list of the actions having occurred from there.

Anyway, I can recompute `prevPlayer` and `autoEndTurn` from `currentTurn` and `lastTurn`.

I'm OK with `autoEndTurn` being recomputed. But should I create a fake `currentTurn` and a fake `lastTurn` when I convert a FEN to a situation, just so that the recomputation of `prevPlayer` from said situation should yield the `prevPlayer` contained in the FEN? That seems insane.

r-gke replied:
Well, in fact `prevPlayer` cannot be computed from `currentTurn` and `lastTurn`. Therefore we must add it to `History` (or to `Uci`, but it does not seem practical).

r-gke replied:
In the last commit, I deleted `LightMove` and `prevMove`, and kept `prevPlayer`.

r-gke replied:
I gave the matter some thought: if it is absolutely necessary, I can retrieve `prevPlayer` from `lastAction`:

1. if empty, no `prevPlayer`;
2. otherwise, compute whether the move was in-line or broadside;
3. if in-line, compute its unit vector and add it to its `orig`, then check the piece there and return the player it belongs to;
4. if broadside, it is simpler: check the piece at `dest` and return the player it belongs to.

Overall, it involves a lot of computations (if we had the previous situation, it would be simpler: just check the piece at `orig`), and cannot deal with an illegal player moving.

msomatt replied:
I would prefer `prevPlayer` not to be stored in `History`. Looks like you are only using `prevPlayer` to write this info into the fen? But we could just write something different into the fen like an integer for how many plies (actions) we've had this turn. i.e. if the fen contained "w 0" we would know it was whites turn and they have two moves to make, if it was "b 1" we would know it was blacks turn and they have already made one move and have one move to make.

Have a look at Amazons (in fairysf) which has a similar turn structure (two actions for every turn). I think for that we store the uci of the turn so far in the fen, which would also work for abalone. Also Monster Chess variant.

The desire to keep History as light-weight as possible is because we have to write all this to the db when saving mid game and we don't want to write some massive tree of historical situations.

Does that all make sense? Can you see work arounds for how to achieve both of our desires?

r-gke replied:
Here it is not exactly like Amazons or Monster, since the number of actions per turn is not always 2: it is 1 for the first turn. Therefore I'd rather have, e.g., the total number of plies than the number of plies of the turn.

Anyway, `prevPlayer` is used to compute the next player (viz. to compute `autoEndTurn`). This is why I write it in the FEN, as well as use it in the hash.

I think the best practice is to compute the next player with local information if possible. Computation via the number of plies or this kind of thing is worse because it relies on global data on the game, and thus assumes the partial validity of all the previous actions (for instance, it makes little sense if the players have not followed the legal order of play).

I don't know how the games are stored in the database, hence I find it difficult to understand the need for `History` there: if I wanted to record a game state, I figure I would only need the initial position and the list of actions up to said state.

r-gke replied:
> ```
> Situation.player | History.currentTurn.isEmpty | prevPlayer
> ===========================================================
>               P1 |                        true |         P2
>               P1 |                       false |         P1
>               P2 |                        true |         P1
>               P2 |                       false |         P2
> ```
> 
> (Technically, this would return P2 at the very start of the game, but i dont think this matters for what you are using it for?)
At the start of the game, it should return P1, because the first turn of P1 has only one action.

r-gke replied:
I think your approach (determining a player from a move) is fundamentally wrong. Imagine I define a variant of Abalone with added neutral pieces that both players can use: no way to guess who moved such pieces if not explicitly recorded. Thus, any technique to find the previous player from a move is a hack that assumes a single player is allowed to perform a given move, which can be false for some (future) variants.
Consider for instance game Yodd (https://boardgamegeek.com/boardgame/105173/yodd), with an order of play similar to Grand Abalone, but wherein each player is free to drop any piece.

By the way, in Amazons you don't need the `Uci` to assess a position, you only need its destination; and in Monster chess, you only need the previous player (unless there is a rule precluding the cancellation of the first action by the second one, which I'm not aware of). Btw, why adding character '½' to the FEN? It seems unnecessary as well.

msomatt replied:
> At the start of the game, it should return P1, because the first turn of P1 has only one action.
Surely by the way you have defined it, prevPlayer should be `None` at the start of the game? Not P1?

I am trying to get as much consistency across similar games. It would seem to make sense to follow the already defined format for multiaction turns from other games. You are correct Amazons only needs to know dest pos, and Monster doesnt need to know the first move at all. It wasn't me who made this decision to include full current turn uci in the FEN (and the ½), so im not going to say its the best thing ever, but it is how things currently work for these games and that's not changing here.

> I think your approach (determining a player from a move) is fundamentally wrong.
I'm not determining a player from a move. We already have games like Othello where placing a piece on a square can be done by either player, so from the uci alone you dont know who made the action. I am trying to explain you can calculate `prevPlayer` from information already available (`Situation.player` and `History.currentTurn.isEmpty`). But I don't think `prevPlayer` is necessary at all. All we want to know is if the next move ends a player's turn. After the first turn this can be deduced by knowing if there was already a move made this turn (the FEN knows this if the uci is included, strategygames can tell this from `History.currentTurn.isEmpty`). For the first turn it is a special case. We have another game with a special case on the first turn (Backgammon, where you cant roll a double on the first turn) - have a look at how that is handled.

Important to note, we are discussing two things here, what the FEN format should be, but also what needs to be stored in the `History` object. And these can be different.

r-gke replied:
> Surely by the way you have defined it, prevPlayer should be None at the start of the game? Not P1?

Being one or the other would not change the computations.

r-gke replied:
> Important to note, we are discussing two things here, what the FEN format should be, but also what needs to be stored in the `History` object. And these can be different.

I still don't understand how you store the games in the database; you told me earlier that changing the FEN would screw up the games in the database, from which I can conclude some FEN strings are written there. If a FEN is enough to describe a situation, why would altering History also have an impact on the database? It means the database use both, which seems to me redundant.

r-gke replied:
> I'm not determining a player from a move. We already have games like Othello where placing a piece on a square can be done by either player, so from the uci alone you dont know who made the action. I am trying to explain you can calculate `prevPlayer` from information already available (`Situation.player` and `History.currentTurn.isEmpty`).

It doesn't make sense: I need `prevPlayer` to determine `Situation.player`. More specifically, if we play move `m`, `m.situationBefore.player == m.situationAfter.player` if and only if `!m.autoEndTurn`, and this is true if and only if `m.situationBefore.prevPlayer != m.situationBefore.player` (assuming `m.situationBefore.prevPlayer == P1` at the start).

r-gke replied:
> I'm not determining a player from a move. We already have games like Othello where placing a piece on a square can be done by either player, so from the uci alone you dont know who made the action. I am trying to explain you can calculate `prevPlayer` from information already available (`Situation.player` and `History.currentTurn.isEmpty`).

I agree, given move `m`, I can determine `m.situationBefore.prevPlayer` from both `m.situationBefore.player` and `m.situationBefore.board.history.currentTurn.isEmpty`; except for the first turn, for which I then have no clue how to do it (I took a look at the BackGammon code, but it didn't help).

Then, in the FEN, what should I write? If I write the content of `currentTurn`, I'm still going to need a 'fake' `Uci` for the start if I want to translate the FEN into a situation, which seems really dirty...

r-gke replied:
Hello again, I just made a few new changes to comply with your desire not to add data to `History`: I removed `prevPlayer` and put in the FEN both `currentTurn` ("-" if empty) and the last element of `lastTurn` ("-" if empty).
From `currentTurn` and `lastTurn` we obtain, as previously said, `autoEndTurn` (and therefore the next player), and `prevPlayer` is only explicitly computed within the hash.

msomatt replied:
Hello, sorry I've read things in a different order to how you intended for them and replied having viewed your most recent changes, without reading these comments.

I also see some of your confusion between the FEN and Game object (and all its related information) in respect to the database. So the database does not store a FEN for a game. It stores everything needed to perfectly recreate the `Game` object, complete with `Situation`, `Board`, and `History`. The `Game` object stores the uci of all the played turns up to this point, so is more than just a FEN.

FENs are very useful for passing around a current Situation (board state + info to be able to continue on from here). FENs are used in many other areas of the site, such as analysis pages, and as an encoding of the Situation.

When playing a Game, we aren't really loading from a FEN at any point, we are just progressing our Game object. So to tell if we are in the first turn or not I think I would look to see if `board.history.lastTurn.isEmpty`. I think this is what you have done, which is great.

In the comments I made on your most recent changes, I advocated for removing lastTurn from the FEN, and now I see why you added it it. However I still think this is redundant information. All we need to know is if fen.player == 'b' (P1) and fen.fullMove == 1 then its the beginning of the game, so the first move ends the turn.

Now when loading out of the FEN in `Forsyth.<<@` we need a place to load into the History to store the fact the first turn has been played. One option is to add a new field to History. I don't like this as its redundant information most of the time (because when playing a game normally it will always just equal `history.lastTurn.isEmpty`), and it is something extra to store in the database. So the other option is to then fake something into lastTurn, even though we don't know it. I know you mentioned this option previously, and it is a bit of a hack, but this route of constructing a game from a FEN isn't the main use-case of the code.  So I think in `Forsyth.<<@` I would do `lastTurn = if (fen.player == Some(P1) && fen.fullMove == 1) List() else List(Uci.Move("a1a1"))`

What do you think to this proposal?

r-gke replied:
> What do you think to this proposal?

I don't think this is an improvement. We'd better delete `halfMoves` and `fullMoves`, which are useless for Abalone, than force ourselves to create fake moves. Moreover, relying on local data (a few previous moves) rather than global data (the count of the turns over the whole game) is more robust.

Neither do I see a good reason to add "½" to the current turn. It is useless and should rather be deleted from Amazons, Monster chess, etc. for consistency.

msomatt replied:
Hi Alex, I've discussed this with the other developers involved in strategygames and we have come to the following conclusion:
- We would like the FEN format to be as follows: `SS2ss/SSS1sss/1SS2ss1/9/ss6SS/sss5SSS/ss6SS/9/1SS2ss1/SSS1sss/SS2ss 0 0 b 0 2 1` where the final integer on the end of the FEN is only present for Grand Abalone and represents `pliesRemainingThisTurn`. So in the initial fen it will be 1 because Black has only 1 move (ply) to make on their first turn. At the end of each turn it is set to 2 as each subsequent turn has 2 moves (plies) and this is decremented with each move/ply.
- The `History` object should store this `pliesRemainingThisTurn` attribute and so this will be the only extra thing to store in the database. It should be an `Option[Int]` and set to `None` for Abalone.
- The Abalone, Amazons and Monster FENs should remain unchanged.

Do you think this will solve all of the problems and be an acceptable solution?

r-gke replied:
> Do you think this will solve all of the problems and be an acceptable solution?

Hi, yes, it is. I have fundamentally nothing against it, as it is almost identical to what I had coded earlier (adding the last player to `History` and to the FEN).

So much so, though, that I find it hard to grasp the reason for it not to be identical... Can you enlighten me?

Likewise, what was wrong with adding nothing in `History` and write one or two `Uci` (according the `Situation`) in the Grand Abalone FEN?

msomatt replied:
Great, if you can implement these changes, then I think we'll be able to look to get this PR accepted.

Part of the reason for the change in stance was due to thinking it over some more, and also after discussions with the other developers as I wasn't satisfied with the hacky solution I proposed, nor was I happy with the changes in FEN format you were proposing. The discussion made me agree the hacky solution was bad, but also that I shouldn't have discounted adding one piece of information into the History. I think it was your initial PR that added multiple fields here that put me off this idea at first.

So apologies if it feels like we have gone round in circles. I agree there is little difference between what you had at one stage with just the `prevPlayer` being in the `History`, although I prefer this solution as its slightly more robust/extendable.

I was happy to have the currentTurn uci written into the FEN as this matched the Monster+Amazons solution, but it seemed very unnecessary to add the lastTurn uci as this doesn't give any more information, as mentioned you could already deduce how many turns were remaining based on `(fen.player == P1 && fen.fullTurnCount == 1)`. Anyway, the minimalist FEN was agreed on by the other developers, so lets go with that.

r-gke replied:
> Great, if you can implement these changes, then I think we'll be able to look to get this PR accepted.

I just made the changes!

r-gke commented on src/main/scala/abalone/Move.scala:
```diff
@@ -1,67 +1,47 @@
+}
 
-  override def toString = s"$piece ${toUci.uci}"
+case class LightMove(
```
I have no issue with deleting `LightMove` and using `Uci` instead. But I think `Uci` should contain the player.

msomatt commented:
Hi @r-gke Sorry for the delayed reply - had some other things on this week. I've given some feedback on the questions you've asked and I'll review your code changes again properly next week.

r-gke commented:
> Hi @r-gke Sorry for the delayed reply - had some other things on this week. I've given some feedback on the questions you've asked and I'll review your code changes again properly next week.

No problem, thanks for your time anyway!

msomatt commented on src/main/scala/abalone/Pos.scala:
```diff
@@ -1,525 +1,382 @@
+
+  val re: Regex = _re.r
+
+  def fromKey(key: String): Option[Pos] = {
```
does this return `None` if the key parameter doesn't match a valid Pos?

r-gke replied:
Why should it?

msomatt replied:
if the input paramater key was to be something like `hello` this function should return None. More likely would be something like `n12` which might be out of range for letter/number

r-gke replied:
Here `hello` returns `None`, but not `n12`, since it represents a point with integral coordinates. Why should `n12` be dismissed?

msomatt replied:
I might be wrong but i thought n12 was outside the range of possible Pos even for Grand Abalone? Perhaps `z99` is a better example?

r-gke replied:
Sure, `"n12"` is outside the board, but I see no reason to dismiss it in `fromKey`. The corresponding `Pos` `p` is dismissed in `BoardType` when checking whether `isCell(p)`.

msomatt reviewed (changes requested):
Hi Alex,

Just had a chance to go through a few more bits.

Sorry if it feels like I'm being really picky and insistent, but I want to keep the style/format the same across the whole of strategygames. So things like `situation` => `sit` make us lose consistency throughout the whole repository, and clutter the diff making it harder to see what you have actually changed.

But hopefully we are nearly there with the strategygames changes :)

r-gke commented on src/main/scala/Pos.scala:
```diff
@@ -126,25 +110,21 @@ object Pos {
+    override def piotr: Char = p.piotr
 
-    def gameLogic: GameLogic = GameLogic.Abalone()
+    override lazy val toInt: Int = (p.x << 3) + p.y
```
Right!

r-gke commented on src/main/scala/abalone/Pos.scala:
```diff
@@ -1,525 +1,382 @@
+  lazy val all: List[Pos] = Piotr.posToIndex.keys.toList
 }
+
+object Piotr {
```
I used the function from `Pos` to character from Go specifically because the `Piotr` previously defined for Abalone was the same, limited to the `Pos` relevant to standard Abalone; it should hence not break anything.

msomatt replied:
Ok, great :)

msomatt commented on src/main/scala/abalone/format/Forsyth.scala:
```diff
@@ -59,15 +59,18 @@ object Forsyth {
+      val lastTurn    = game.situation.board.history.lastTurn.reverse.headOption
+      val currentTurn = game.situation.board.history.currentTurn.reverse.headOption
+
+      " " + currentTurn.fold("-")(_.uci) + " " + lastTurn.fold("-")(_.uci)
```
To keep consistency, we just want currentTurn to be put in after the `½` character. 
We also don't need a '-' if no currentTurn.
Something like
```
val currentTurn = game.situation.board.history.currentTurn.map(_.uci).mkString(",")
if (currentTurn.size > 0) s" ½$currentTurn" else ""
```

msomatt commented on src/main/scala/abalone/format/FEN.scala:
```diff
@@ -30,32 +30,36 @@ final case class FEN(value: String) extends AnyVal {
 
-  def fullMove(variant: Variant): Option[Int] = intFromFen(if (variant.hasPrevPlayer) 6 else 5)
+  def fullMove: Option[Int] = intFromFen(5)
+
```
I dont think we need to put lastTurn into the FEN?

To keep consistency, we just want currentTurn to be put in after the ½ character.

This way we don't even need to pass through the variant we can just see is there is a ½ in the fen and build up the currentTurn from that. I prefer not needing to pass the variant into functions in the FEN class.

Amazons does something like:
`Uci.Move.apply(variant.gameFamily, fen.value.dropWhile('½' !=).drop(1))` which might also work here.

msomatt commented on src/main/scala/abalone/format/Forsyth.scala:
```diff
@@ -10,22 +10,22 @@ import strategygames.{Player, Score}
           pieces = fen.pieces(variant),
           history = History(
-            prevPlayer = fen.prevPlayer(variant),
+            lastTurn = fen.lastTurn(variant).fold(List[Uci]())(List(_)),
```
Yeah so we don't need to rebuild the lastTurn from the FEN so I would omit this. Its ok not to have this information in the FEN

msomatt commented on src/main/scala/abalone/variant/GrandAbalone.scala:
```diff
@@ -34,10 +34,10 @@ case object GrandAbalone
   /** Belgian daisy. */
   override def initialFen: FEN =
-    format.FEN("SS2ss/SSS1sss/1SS2ss1/9/ss6SS/sss5SSS/ss6SS/9/1SS2ss1/SSS1sss/SS2ss 0 0 b * 0 1")
+    format.FEN("SS2ss/SSS1sss/1SS2ss1/9/ss6SS/sss5SSS/ss6SS/9/1SS2ss1/SSS1sss/SS2ss 0 0 b 0 1 - -")
```
Do the pieces need to be flipped as your FEN format changed (like you just did with the Abalone `initialFen`?
I would remove the final two `- -` here.

msomatt reviewed (changes requested):
Thanks for resetting `sit`. Just some tweaks with the FEN format now, but we're nearly there I think!

msomatt commented:
Hi @r-gke I've merged your branch into our repo, but I didn't accept it into `dev` branch as that would mean that this version is ready to be used on our dev/live environments, which it isn't.
I built the new strategygames with your changes and tried to run it with https://github.com/Mind-Sports-Games/lila in the PS website. There were a few small changes that were needed in lila to get things compiling as the changes alter the way the BoardSize is handled, and Pos as well. I was able to get past these for compilation but the changes in Pos mean you cant successfully play an Abalone game. All understandable, but I think that's the next task for you - getting Abalone fully working. Let us know if you need help setting up lila/lila-ws etc

r-gke commented:
> Hi @r-gke I've merged your branch into our repo, but I didn't accept it into `dev` branch as that would mean that this version is ready to be used on our dev/live environments, which it isn't. I built the new strategygames with your changes and tried to run it with https://github.com/Mind-Sports-Games/lila in the PS website. There were a few small changes that were needed in lila to get things compiling as the changes alter the way the BoardSize is handled, and Pos as well. I was able to get past these for compilation but the changes in Pos mean you cant successfully play an Abalone game. All understandable, but I think that's the next task for you - getting Abalone fully working. Let us know if you need help setting up lila/lila-ws etc

I have a fork for lila, but my IDE seems unable to identify the project structure. Thus I'm basically stuck with a text editor, which makes it very difficult to spot what I have to modify for it to work. Moreover, I've been unable to make lila/lila-ws work (or anything remotely close to it), so that I cannot test what I do.
The best help for me would be to indicate me which files I have to edit.

By the way, I also have a fork of chessground, for which I rewrote the typescript code to match what I introduced in playstrategy. Amongst other things, I rewrote there the code to translate a Pos into a drawn point, but of course I did not test my functions to check whether they were correct.

msomatt commented:
Hi @r-gke I've spoken with @vincentfrochot and he will assist you with setting up lila and making changes in chessground.

With regards to your Pos changes, remember that all other game logics (Chess, Draughts, Go, Backgammon etc etc) will be using an unaltered style, so therefore any changes in lila, chessground will need to work with all the games in the PlayStrategy library. We won't want to make a such a huge change to how Pos work for all games in releasing Grand Abalone. If you are having trouble with the your new Pos class it perhaps needs to be converted into a more similar structure to the rest of the Pos structures in the 'wrapper' layer in strategygames (https://github.com/Mind-Sports-Games/strategygames/blob/dev/src/main/scala/Pos.scala)
